### ISSUE TITLE
Hyperparameters when training Cross-Encoder.

### ISSUE THREAD

**Issue by huhk-sysu** [chijames/Poly-Encoder#2]
Hi! I'm using your code and want to reproduce your result on the DSTC 7 dataset.
When training the Cross-Encoder. I use BERT-small (uncased_L-4_H-512_A-8.zip) and leave all hyperparameters unchanged as in run.py (batch size=32, max context length=128, max response length=32). However I came across OOM on my Tesla M40 GPU, which has a memory of 11G.
I wonder how you can train the cross-encoder on your GPU. I guess the default hyperparameters in run.py are designed for training bi-encoder and poly-encoder. Could you please show me your hyperparameters when training cross-encoder?

**Comment by huhk-sysu** [comment]
I also found your results on Poly-Encoder are strange. It seems that on both datasets, your Poly-Encoder with a code size of 16 behaved better than those who has a larger size like 64 and even 360. It conflicts with the results in the paper and [sfzhou5678](https://github.com/sfzhou5678/PolyEncoder), where the scores consistently improve as the code size become larger. Is there any explanation?

**Comment by chijames** [comment]
Hi,

Please reduce the batch size to what fit in your GPU, that's what I did on my machine. I leave other parameters unchanged.

As for performance, I think this is reasonable. First of all, if you take a closer look at the last table (Table 10) of the original [paper](https://arxiv.org/pdf/1905.01969.pdf), you can see that the performance does NOT improve consistently as poly_m becomes larger. Secondly, I only report the result of one run, it's better if the numbers are averaged across multiple runs. The original paper also reports the +- range in the table. Lastly, some metrics like R@2, R@5, MRR do improve on the ubuntu dataset, and they are R@1 and R@2 on the dstc7 dataset.

I also want to highlight that the original paper uses the data augmentation technique and a larger model. I am pretty sure that the first factor greatly affects the performance (improve stability, prevent overfitting, and so on). If you have time and resources, you can try to run the experiments using the augmented data (and for more epochs) as I put in the readme.

With that being said, it's very likely that there are bugs in my implementation. Feel free to open another issue if you spot anything weird.

Thanks.

**Comment by huhk-sysu** [comment]
Thanks for your response! I carefully check the Table 10 and find you are right. Maybe different datasets or pre-trained models prefer different code size.

I haved tried the augmented data, keep all parameters unchanged but train for 3 epochs like sfzhou5678. In recent experiments, the improvement is not obvious enough (poly-encoder-16's MRR only improve 0.5 from bi-encoder, and poly-encoder-64 improve 0.2 from poly-encoder-16). I attributed this phenomenon to the random factors and maybe it needs multiple runs and longer training time to avoid randomness.

Since I'm new to this area, I also want to know how to "cache the pre-computed responses and utilize them to accelerate inferring" with your code. Is there any example?

By the way, I have checked your implementation and didn't find any bugs, at least for me.  Thanks for your awesome work.

**Comment by chijames** [comment]
Regarding "how to cache pre-computed...":

TL;DR: My implementation does not support this (and I believe the original ParlAI release does not either).

Long answer:
In the setting adopted by the original paper, every test instance is unique (different context, candidate responses). However, in a real world/production setting, given a context, we want to retrieve the best response in our database. Here, "database" is just a pile of responses, you can treat them as the training data (all the 100 candidate responses) in, for example, dstc7. In other words, only contexts are changing but responses are not (they come from the "database").

Note we also have the trained model on dstc7 in hand. If we are using bi/poly encoders, we can precompute and cache bert sentence embeddings for each response in the "database". Given those cache vectors of responses, during testing time, we just need to encode contexts and calculate attentions with poly vectors and similarity scores for ranking. This is not possible for the cross encoder, as it needs to concatenate a context with each possible response in the very beginning to form vector representations. For a better understanding, take a look at the original paper's figure.

Thanks.
